### PR TITLE
i#2985 drx_expand_scatter_gather(): Add a test skeleton for scatter + gather.

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds the test drxmgr_scattergather-test. Currently, the test does nothing but running AVX2
and AVX-512 scatter + gather sequences and checking for correctness. The test will get
extended in the future to include the future emulation drx emulation sequence expansion
function drx_expand_scatter_gather().

Issue: #2985

### CODE REVIEW

**Comment by hgreving2304** [comment]
> AppVeyor build failed
I will fix the Windows build failure when I have a machine available. __u32? OUT?

**Comment by hgreving2304** [comment]
> Travis CI - Pull Request Failing after 32m — Build Failed

Due to one of the AVX tests using AVX2 instructions. We need another check for AVX2, on top of the existing AVX and AVX-512 check.

**Comment by hgreving2304** [comment]
> > Travis CI - Pull Request Failing after 32m — Build Failed
> 
> Due to one of the AVX tests using a vpbroadcast version that encodes to a AVX-512 instruction. Will fix.

Done.

File: core/arch/asm_defines.asm [hunk 1]
```diff
@@ -328,8 +328,17 @@ ASSUME fs:_DATA @N@\
 #   define LIB_SEG_TLS gs /* keep in sync w/ unix/os_exports.h defines */
 #  endif
 # endif /* 64/32-bit */
+#  define REG_XMM0 xmm0
```
Comment (derekbruening) [inline core/arch/asm_defines.asm]: There's no other "xmm" so I'm not sure why we need defines like these?  The REG_XAX, etc. are to hide the rax/eax distinction.  There's no variation here.  Wouldn't it be better to just use "xmm0" in the asm?

File: suite/tests/CMakeLists.txt [hunk 2]
```diff
@@ -2213,6 +2213,20 @@ if (CLIENT_INTERFACE)
   use_DynamoRIO_extension(client.drxmgr-test.dll drx)
   use_DynamoRIO_extension(client.drxmgr-test.dll drreg)
 
+  if (X86)
+    tobuild_ci(client.drxmgr_scattergather-test client-interface/drxmgr_scattergather-test.c "" "" "")
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Not targeted towards drmgr, but towards drx. The future drx_expand_scatter_gather() will be part of it.

[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Renamed. I left drx because of above. WDYT?

File: core/arch/asm_defines.asm (hunk 1 above)
Comment (hgreving2304) [reply core/arch/asm_defines.asm]: Ok. Renamed to regular names. Previously, REG_ZMM versions had been added. We then might want to remove them as well.

File: suite/tests/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply suite/tests/CMakeLists.txt]: I could see it either way: such a test might include core DR VSIB operand iteration tests, e.g. and not just be a drx test.

**Comment by hgreving2304** [comment]
Merged, assuming you hadn't looked at it yet.

File: suite/tests/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply suite/tests/CMakeLists.txt]: I have to say that I find the name to be inconsistent: the mix of underscore and hyphen bothers me: "drx_scattergather-test"

[the remark this answers is not in the document]
Comment (derekbruening) [reply suite/tests/CMakeLists.txt]: I would just s/-test// altogether.

[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: I took it over from client.drx_buf-test. Happy to rename, into..?

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+    }
+    return true;
+}
+
+static bool
+test_avx2_avx512_scatter_gather(void)
+{
+#    if defined(__AVX512F__) || defined(__AVX__)
+    uint32_t ref_sparse_test_buf[] = {
+        0x0, 0x1, 0xf, 0xf, 0x1, 0x2, 0xf, 0xf, 0x2, 0x3, 0xf, 0xf, 0x3, 0x4, 0xf, 0xf,
```
Comment (derekbruening) [inline suite/tests/client-interface/drx_scattergather-test.c]: Please add a comment explaining where this pattern came from and why it was chosen.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 4]
```diff
@@ -0,0 +1,772 @@
...
+    uint32_t ref_idx32_val32_xmm_ymm_zmm[] = { /* xmm */
+                                               0x0, 0x1, 0x2, 0x3,
+                                               /* ymm */
+                                               0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7,
+                                               /* zmm */
+                                               0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7,
+                                               0x8, 0x9, 0xa, 0xb, 0xc, 0xd, 0xe, 0xf
+    };
+    uint32_t ref_idx32_val64_xmm_ymm_zmm[] = { /* xmm */
+                                               0x0, 0x1, 0x1, 0x2,
```
Comment (derekbruening) [inline suite/tests/client-interface/drx_scattergather-test.c]: Same with all of these: not understanding these sequences.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 5]
```diff
@@ -0,0 +1,772 @@
...
+        ret                                                @N@\
+        END_FUNC(funcname)
+
+#undef OPCODE
+#define OPCODE vpscatterqq
+TEST_AVX512_SCATTER_IDX64_VAL64(FUNCNAME(OPCODE), OPCODE)
+#undef OPCODE
+#define OPCODE vscatterqpd
+TEST_AVX512_SCATTER_IDX64_VAL64(FUNCNAME(OPCODE), OPCODE)
+#endif /* __AVX512F__ */
```
Comment (derekbruening) [inline suite/tests/client-interface/drx_scattergather-test.c]: suggest: banner comment to help identify these different regions: avx512 vs avx2

File: suite/tests/CMakeLists.txt [hunk 6]
```diff
@@ -2213,6 +2213,23 @@ if (CLIENT_INTERFACE)
   use_DynamoRIO_extension(client.drxmgr-test.dll drx)
   use_DynamoRIO_extension(client.drxmgr-test.dll drreg)
 
+  if (X86)
+    if (NOT MACOS)
+      # XXX i#2985: The test's asm doesn't build with MacOS's clang.
+      tobuild_ci(client.drx_scattergather-test client-interface/drx_scattergather-test.c "" "" "")
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Renamed into drx-scattergather.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 7]
```diff
@@ -0,0 +1,772 @@
...
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* TODO i#2985: Currently, this test doesn't do anything but running scatter/gather
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 8]
```diff
@@ -0,0 +1,772 @@
...
+ * sequences in AVX-512 and AVX2 and checking for correctness. This test will get extended
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 9]
```diff
@@ -0,0 +1,772 @@
...
+        kmovw      k1, edx                                @N@\
+        opcode     zmm0 {k1}, [REG_XAX + zmm1 * 4]        @N@\
+        vmovdqu32  [REG_XCX + 48], zmm0                   @N@\
+        add        REG_XSP, FRAME_PADDING                 @N@\
+        POP_CALLEE_SAVED_REGS()                           @N@\
+        ret                                               @N@\
+        END_FUNC(funcname)
+
+#undef OPCODE
+#define OPCODE vpgatherdd
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 10]
```diff
@@ -0,0 +1,772 @@
...
+#else
+# define FRAME_PADDING 0
+#endif
+
+#define JOIN(x, y) x ## y
+#define FUNCNAME(opcode) JOIN(test_avx512_, opcode)
+
+#ifdef __AVX512F__
+#define TEST_AVX512_GATHER_IDX32_VAL32(funcname, opcode)  @N@\
+DECLARE_FUNC_SEH(funcname)                                @N@\
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 11]
```diff
@@ -0,0 +1,772 @@
...
+void
+test_avx2_vgatherqpd(uint32_t *ref_sparse_test_buf, uint32_t *test_idx32_vec,
+                     uint32_t *output_xmm_ymm_zmm OUT);
+
+#    define SPARSE_FACTOR 4
+#    define XMM_REG_SIZE 16
+#    define YMM_REG_SIZE 32
+#    define ZMM_REG_SIZE 64
+#    define CONCAT_XMM_YMM_ZMM_U32 \
+        (XMM_REG_SIZE + YMM_REG_SIZE + ZMM_REG_SIZE) / sizeof(uint32_t)
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c (hunk 5 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 12]
```diff
@@ -0,0 +1,772 @@
...
+#ifndef ASM_CODE_ONLY /* C code */
+#    include <stdint.h>
+#    include "tools.h"
+
+#    ifndef X86
+#        error "This test is x86 specific."
+#    endif
+
+void
+test_avx512_vpscatterdd(uint32_t *xmm_ymm_zmm, uint32_t *test_idx32_vec,
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Added comments.

File: suite/tests/client-interface/drx_scattergather-test.c [hunk 13]
```diff
@@ -0,0 +1,772 @@
...
+/* See comment above. */
+void
+test_avx2_vgatherqpd(uint32_t *ref_sparse_test_buf, uint32_t *test_idx32_vec,
+                     uint32_t *output_xmm_ymm_zmm OUT);
+
+#    define SPARSE_FACTOR 4
+#    define XMM_REG_SIZE 16
+#    define YMM_REG_SIZE 32
+#    define ZMM_REG_SIZE 64
+#    define CONCAT_XMM_YMM_ZMM_U32 \
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Added comments.

File: suite/tests/client-interface/drx_scattergather-test.c (hunk 3 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Done.

File: suite/tests/client-interface/drx_scattergather-test.c (hunk 4 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Added comments.

File: suite/tests/client-interface/drx_scattergather-test.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx_scattergather-test.c]: Leaving this unresolved to cover comments in general.

File: suite/tests/client-interface/drx-scattergather.c [hunk 14]
```diff
@@ -204,12 +214,35 @@ static bool
...
+        0x9, 0xa, 0xf /* poison */, 0xf /* poison */,
+        0xa, 0xb, 0xf /* poison */, 0xf /* poison */,
+        0xb, 0xc, 0xf /* poison */, 0xf /* poison */,
+        0xc, 0xd, 0xf /* poison */, 0xf /* poison */,
+        0xd, 0xe, 0xf /* poison */, 0xf /* poison */,
+        0xe, 0xf, 0xf /* poison */, 0xf /* poison */,
+        0xf, 0x0, 0xf /* poison */, 0xf /* poison */
     };
+    /* As pointed out above, the ref_xmm_ymm_zmm are the concatenated results (gather
+     * tests) or sources (scatter tests) for the xmm, ymm, and zmm version of the gather
```
Comment (derekbruening) [inline suite/tests/client-interface/drx-scattergather.c]: s/version/versions/

File: suite/tests/client-interface/drx-scattergather.c [hunk 15]
```diff
@@ -252,6 +285,10 @@ test_avx2_avx512_scatter_gather(void)
     uint32_t output_sparse_test_buf[SPARSE_TEST_BUF_SIZE_U32];
 #    endif
 #    ifdef __AVX512F__
+    /* As pointed out above, the gather tests gather data from ref_sparse_test_buf, and
+     * concatenate the results of each xmm, ymm and zmm version of the gather instruction
+     * in output_xmm_ymm_zmm. The output is expected to be the ref_xmm_ymm_zmm buffer
```
Comment (derekbruening) [inline suite/tests/client-interface/drx-scattergather.c]: nit: missing trailing .

File: suite/tests/client-interface/drx-scattergather.c [hunk 16]
```diff
@@ -284,6 +321,11 @@ test_avx2_avx512_scatter_gather(void)
                             ref_idx64_val64_xmm_ymm_zmm, test_idx64_vec,
                             output_xmm_ymm_zmm))
         return false;
+    /* As pointed out above, the scatter tests scatter data from ref_xmm_ymm_zmm into the
+     * array output_sparse_test_buf. Its the inverse of the gather test, so the source
```
Comment (derekbruening) [inline suite/tests/client-interface/drx-scattergather.c]: grammar: s/Its/It's/

File: suite/tests/client-interface/drx-scattergather.c [hunk 17]
```diff
@@ -284,6 +321,11 @@ test_avx2_avx512_scatter_gather(void)
...
+     * data for the xmm, ymm, and zmm scatter instruction is concatenated in
```
Comment (derekbruening) [inline suite/tests/client-interface/drx-scattergather.c]: s/instruction is/instructions are/ OR s/the/each/

File: suite/tests/client-interface/drx-scattergather.c [hunk 18]
```diff
@@ -30,9 +30,19 @@
...
+/* TODO i#2985: Currently, this test doesn't do anything but run scatter/gather sequences
+ * in AVX-512 and AVX2 and check for correctness. This test will get extended to include
+ * the future drx_expand_scatter_gather() extension.
+ */
+
+/* The test gathers data from a buffer called sparse_test_buf, runs the xmm, ymm, and zmm
+ * version of the gather instruction, and concatenates the results of each version into a
+ * new buffer called xmm_ymm_zmm.
+ *
+ * Similarly, the scatter tests do the inverse and scatters the xmm, ymm, and zmm data
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c [hunk 19]
```diff
@@ -30,9 +30,19 @@
...
 
-/* TODO i#2985: Currently, this test doesn't do anything but running scatter/gather
- * sequences in AVX-512 and AVX2 and checking for correctness. This test will get extended
- * to include the future drx_expand_scatter_gather() extension.
+/* TODO i#2985: Currently, this test doesn't do anything but run scatter/gather sequences
+ * in AVX-512 and AVX2 and check for correctness. This test will get extended to include
+ * the future drx_expand_scatter_gather() extension.
+ */
+
+/* The test gathers data from a buffer called sparse_test_buf, runs the xmm, ymm, and zmm
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Done.

File: suite/tests/client-interface/drx-scattergather.c [hunk 20]
```diff
@@ -30,9 +30,19 @@
...
+ * version of the gather instruction, and concatenates the results of each version into a
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c [hunk 21]
```diff
@@ -204,12 +214,35 @@ static bool
 test_avx2_avx512_scatter_gather(void)
 {
 #    if defined(__AVX512F__) || defined(__AVX__)
+    /* The data in the uint32_t sparse array is layed out in 2 uint32_t, followed by 2
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c [hunk 22]
```diff
@@ -204,12 +214,35 @@ static bool
...
+     * poisen 0xf. The values are randomly chosen to be N = 0, N+1, 0xf, 0xf, N+1, N+2,
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c [hunk 23]
```diff
@@ -204,12 +214,35 @@ static bool
...
+     * ... The dword value scatter/gather instructions write/read 1 uint32_t, while the
+     * qword scatter/gather tests write/read both uint32_t. The are maximal 16 values
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c [hunk 24]
```diff
@@ -204,12 +214,35 @@ static bool
...
+     * being scattered/gathered. So the array is 16x4 uint32_t long.
+     */
     uint32_t ref_sparse_test_buf[] = {
-        0x0, 0x1, 0xf, 0xf, 0x1, 0x2, 0xf, 0xf, 0x2, 0x3, 0xf, 0xf, 0x3, 0x4, 0xf, 0xf,
-        0x4, 0x5, 0xf, 0xf, 0x5, 0x6, 0xf, 0xf, 0x6, 0x7, 0xf, 0xf, 0x7, 0x8, 0xf, 0xf,
-        0x8, 0x9, 0xf, 0xf, 0x9, 0xa, 0xf, 0xf, 0xa, 0xb, 0xf, 0xf, 0xb, 0xc, 0xf, 0xf,
-        0xc, 0xd, 0xf, 0xf, 0xd, 0xe, 0xf, 0xf, 0xe, 0xf, 0xf, 0xf, 0xf, 0x0, 0xf, 0xf
+        0x0, 0x1, 0xf /* poison */, 0xf /* poison */,
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Done.

File: suite/tests/client-interface/drx-scattergather.c (hunk 14 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c (hunk 15 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c (hunk 16 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Fixed.

File: suite/tests/client-interface/drx-scattergather.c (hunk 17 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drx-scattergather.c]: Done.

**Comment by derekbruening** [comment]
A reminder to please read through thoroughly before committing -- there were just too many writing errors.

**Pull request merged**